### PR TITLE
DnsNameResolver should not bind locally. Fixes #5457

### LINKED ISSUE #5457: DnsNameResolver should not bind locally
Netty version: 4.1.1.Final (and master)

Context:
Dns resolution failures when using the `DnsNameResolver` and the JVM is not authorized to bind datagram channels.

Steps to reproduce:
Use the `DnsNameResolver` with an environment that does not allow to bind datagram channels.

The current `DnsNameResolver ` binds locally a `DatagramChannel` which is not necessary (and not always permitted).

The bind is used to obtain a channel and it is possible to create a `DatagramChannel` directly without the need to bind. The method `newChannel` in `DnsNameResolver` can be changed from:

```
  private ChannelFuture newChannel(
      ChannelFactory<? extends DatagramChannel> channelFactory, InetSocketAddress localAddress) {

    Bootstrap b = new Bootstrap();
    b.group(executor());
    b.channelFactory(channelFactory);
    final DnsResponseHandler responseHandler = new DnsResponseHandler();
    b.handler(new ChannelInitializer<DatagramChannel>() {
      @Override
      protected void initChannel(DatagramChannel ch) throws Exception {
        ch.pipeline().addLast(DECODER, ENCODER, responseHandler);
      }
    });

    ChannelFuture bindFuture = b.bind(localAddress);
    bindFuture.channel().closeFuture().addListener(new ChannelFutureListener() {
      @Override
      public void operationComplete(ChannelFuture future) throws Exception {
        resolveCache.clear();
      }
    });

    return bindFuture;
  }
```

to

```
  private ChannelFuture newChannel(
      ChannelFactory<? extends DatagramChannel> channelFactory, InetSocketAddress localAddress) {

    DatagramChannel channel = channelFactory.newChannel();
    channel.config().setOption(ChannelOption.DATAGRAM_CHANNEL_ACTIVE_ON_REGISTRATION, true);
    channel.config().setRecvByteBufAllocator(new FixedRecvByteBufAllocator(maxPayloadSize));
    final DnsResponseHandler responseHandler = new DnsResponseHandler();
    channel.pipeline().addLast(DECODER, ENCODER, responseHandler);
    channel.closeFuture().addListener(new ChannelFutureListener() {
      @Override
      public void operationComplete(ChannelFuture future) throws Exception {
        resolveCache.clear();
      }
    });

    return executor().register(channel);
  }
```

to achieve the same without binding.

### PR DESCRIPTION
**Pull request by vietj**
Motivation:
Dns resolution failures happen when using the DnsNameResolver and the JVM is not authorized to bind datagram channels. The current DnsNameResolver binds locally a DatagramChannel which is not necessary (and not always permitted).

Modifications:
The DnsNameResolver Bootstrap does not bind anymore, instead it registers and use the channel directly. The localAddress has also been removed from the DnsAddressResolverGroup, DnsNameResolver and DnsNameResolverBuilder as it is not necessary anymore and the API is marked as @UnstableApi.

Result:
Dns resolution does not require anymore to bind locally.

### CODE REVIEW

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java [hunk 1]
```diff
@@ -151,6 +151,8 @@ public void operationComplete(Future<AddressedEnvelope<DnsResponse, InetSocketAd
                         if (traceEnabled) {
                             addTrace(future.cause());
                         }
+                        System.out.println("----------------------- " + question.name());
+                        future.cause().printStackTrace();
```
Comment (slandelle) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java]: woups

**Comment by vietj** [comment]
yeah :-)

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java [hunk 2]
```diff
@@ -606,6 +597,13 @@ private InetSocketAddress nextNameServerAddress() {
     }
 
     private final class DnsResponseHandler extends ChannelInboundHandlerAdapter {
+
+        private final Promise<Channel> channelActivePromise;
+
+        public DnsResponseHandler(Promise<Channel> channelActivePromise) {
```
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java]: @vietj make the constructor package private

**Comment by normanmaurer** [comment]
@vietj LGTM... will pull in once ci job finished

**Comment by normanmaurer** [comment]
@vietj I squashed it and cherry-picked.

4.1 (804e058e27e16e346b565ce4490f6177574e267d)

**Pull request closed without merging**